[ BDH-235 ] Bundler v2

blewater opened the pull request:
- Componentized Intent Supporting Bundler for accessible downstream updates.
- `sendUserOperation` requests are synchronous with the final on-chain submission trx hash.
- Return 3 hashes, original, solved userOp Hash, and Trx hash on `sendUserOperation` requests.
- Passing the `solved` userOpHash to `eth_getUserOperationReceipt` returns the on-chain userOp trx.

blewater committed:
chore: gitignore idea

blewater committed:
chore: gitignore .env

blewater committed:
feat: compatible geth go.mod

blewater committed:
feat: Initial middleware

blewater committed:
feat: Add configuration

blewater committed:
feat: Add main

blewater committed:
refactor: Move env vars to own package

blewater committed:
feat: Extend RPC handlers

blewater committed:
feat: Add client validations

blewater committed:
feat: handle ethSendUserOperation

blewater committed:
feat: Upg Model 18.1

blewater committed:
feat: Communicate hashes to client

blewater committed:
feat: Integrate Solver

blewater committed:
fix: Allow body to be read again by embedded handler

blewater committed:
refactor: Rename and replace Success string with bool

blewater committed:
feat: Remove custom Json marshaling

blewater commented on rpc/main.go:
```diff
@@ -0,0 +1,274 @@
+	"github.com/blndgs/bundler/validations"
+)
+
+func main() {
```
Can move to a `start` folder and rename the package here to `rpc`

blewater replied:
func `main()` does not need to be 200 lines. Can be split.

blewater commented on rpc/handler.go:
```diff
@@ -0,0 +1,383 @@
+	return func(c *gin.Context) {
+		if c.Request.Method != "POST" {
+			jsonrpcError(c, -32700, "Parse error", "POST method excepted", nil)
+			return
+		}
+
+		if c.Request.Body == nil {
+			jsonrpcError(c, -32700, "Parse error", "No POST data", nil)
+			return
+		}
+
+		body, err := readBody(c)
+		if err != nil {
+			jsonrpcError(c, -32700, "Parse error", "Error while reading request body", nil)
+			return
+		}
+
+		data := make(map[string]any)
+		err = json.Unmarshal(body, &data)
+		if err != nil {
+			jsonrpcError(c, -32700, "Parse error", "Error parsing json request", nil)
+			return
+		}
+
+		id, ok := parseRequestId(data)
+		if !ok {
+			jsonrpcError(c, -32600, "Invalid Request", "No or invalid 'id' in request", nil)
+			return
+		}
+
+		if data["jsonrpc"] != "2.0" {
+			jsonrpcError(c, -32600, "Invalid Request", "Version of jsonrpc is not 2.0", &id)
+			return
+		}
+
+		method, ok := data["method"].(string)
+		if !ok {
+			jsonrpcError(c, -32600, "Invalid Request", "No or invalid 'method' in request", &id)
+			return
+		}
```
this boilerplate code runs twice for rpc methods handled by the stackup controller. If possible, it would be nice to run once only.

blewater commented on validations/userop.go:
```diff
@@ -0,0 +1,92 @@
+		g.Go(func() error { return checks.ValidateVerificationGas(ctx.UserOp, v.ov, v.maxVerificationGas) })
+		g.Go(func() error { return checks.ValidatePaymasterAndData(ctx.UserOp, ctx.GetPaymasterDepositInfo(), gc) })
+		g.Go(func() error { return checks.ValidateCallGasLimit(ctx.UserOp, v.ov) })
+		// g.Go(func() error { return checks.ValidateFeePerGas(ctx.UserOp, gasprice.GetBaseFeeWithEthClient(v.eth)) })
```
All the code in this file is copied from stackup to avoid running the above function.

blewater commented on go.mod:
```diff
@@ -0,0 +1,105 @@
+
+require (
+	github.com/dgraph-io/badger/v3 v3.2103.5
+	github.com/ethereum/go-ethereum v1.11.5
```
for the time being, need to pin to `Geth 1.11` but maybe not 1.11.5

blewater commented on rpc/main.go:
```diff
@@ -0,0 +1,274 @@
+func main() {
+	values := conf.GetValues()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```
Maybe not needed because it takes hitting `ctrl-c' twice to exit.

adelowo replied:
fixed this

blewater commented on rpc/main.go:
```diff
@@ -0,0 +1,274 @@
+		log.Fatal(err)
+	}
+
+	c := client.New(mem, gas.NewDefaultOverhead(), chain, values.SupportedEntryPoints, values.OpLookupLimit)
```
the Bundler's front end coupled with the gin web server.

blewater commented on rpc/main.go:
```diff
@@ -0,0 +1,274 @@
+		rep.IncOpsSeen(),
+	)
+
+	b := bundler.New(mem, chain, values.SupportedEntryPoints)
+	b.SetGetBaseFeeFunc(gasprice.GetBaseFeeWithEthClient(eth))
+	b.SetGetGasTipFunc(gasprice.GetGasTipWithEthClient(eth))
```
The bundler's backend.

blewater commented on rpc/main.go:
```diff
@@ -0,0 +1,274 @@
+	log.Println("Shutting down...")
+}
+
+func NewLogger() zerologr.Logger {
+	zl := zerolog.New(os.Stderr).With().Caller().Timestamp().Logger()
+	logger := zerologr.New(&zl)
+	return logger
+}
+
+func runDBGarbageCollection(db *badger.DB) {
+	go func(db *badger.DB) {
+		ticker := time.NewTicker(5 * time.Minute)
+		defer ticker.Stop()
+
+		for range ticker.C {
+		again:
+			err := db.RunValueLogGC(0.7)
+			if err == nil {
+				goto again
+			}
+		}
+	}(db)
+}
```
This and the next function do not belong in `main.go`

blewater commented on rpc/utils.go:
```diff
@@ -0,0 +1,20 @@
+
+// GetClientIPFromXFF returns the client ID using x-forwarded-for headers before relying on c.ClientIP().
+// This assumes use of a trusted proxy.
+func GetClientIPFromXFF(c *gin.Context) string {
```
can be moved or it's duplicate.

blewater commented on solution/solveintents.go:
```diff
@@ -0,0 +1,216 @@
+	}
+}
+
+func ReportSolverHealth(solverURL string) error {
```
It is not an immediate priority, but in the future, it may run once every 20 minutes or so since the last interaction.

blewater reviewed (commented):
Educational comments. We should strive to deploy soon.

hpsing commented on solution/solveintents.go:
```diff
@@ -0,0 +1,216 @@
+
+		// cast the received userOp batch to a slice of model.UserOperation
+		// to be sent to the Solver
+		modelUserOps := *(*[]*model.UserOperation)(unsafe.Pointer(&ctx.Batch))
```
`unsafe.Pointer` to cast ctx.Batch to a []*model.UserOperation. it's risky because it bypasses Go type safety.

blewater replied:
Sure, I moved the compiler safety check for the cast right before it. The alternative of repeatedly creating the model objects for every solver request to copy values will put unnecessary GC pressure and unnecessary overhead.

adelowo replied:
fixed in https://github.com/blndgs/bundler/pull/1/commits/c1f8e4e6661105d8ae1dd8a9fe823c26676f764a

hpsing commented on solution/solveintents.go:
```diff
@@ -0,0 +1,216 @@
+		return err
+	}
+	defer resp.Body.Close()
+
```
```suggestion
if resp.StatusCode != http.StatusOK {
    // Read the response body for error details if necessary
    return fmt.Errorf("Solver returned non-OK status: %s", resp.Status)
}
```

hpsing replied:
solver can give non success response too.

adelowo replied:
fixed in https://github.com/blndgs/bundler/pull/1/commits/c1f8e4e6661105d8ae1dd8a9fe823c26676f764a

blewater committed:
doc: PR feedback

adelowo commented on o11y/init.go:
```diff
@@ -0,0 +1,113 @@
+	return resources
+}
+
+func IsEnabled(serviceName string) bool {
+	return len(serviceName) > 0
+}
```
Should we have a config value strictly for enabling telemetry?

adelowo replied:
ignore. fixed in PR

adelowo commented on rpc/handler.go:
```diff
@@ -0,0 +1,383 @@
+
+func waitForUserOpCompletion(ctx context.Context, ethClient *ethclient.Client, txHashes *xsync.MapOf[string, srv.OpHashes], userOpHash common.Hash) (*HashesResponse, error) {
+	const (
+		waitTimeout = 300 * time.Second
```
we should move this to a config value probably?

adelowo replied:
ignore. fixed in PR

adelowo reviewed (commented):
Looks good to me. I have created a PR at https://github.com/blndgs/bundler/pull/2 that incorporates the feedback on this PR and what we spoke about in https://github.com/blndgs/bundler/pull/2

Tested OTEL, bundling and on chain txs already. 

I think there should be two larger tasks from this as a next step for myself:

- Health checks should run periodically and we probably should expose a bundler  health check endpoint for clients too that takes into account the last known status of the solver ( and maybe future reps )
- We need to standardise logging and add more of them. We are just returning the errors to client and no logging on our end

adelowo committed:
[ BDH-235 ] Bundler v2 updates (#2)

* fix CTRL+C not actually closing the process

* fix otel collector. Set default srvice name and create new config

* use config value to determine wait timeout

* use context timeout cancellation as time#After does not garbage collect

* remove unnneeded repetitions

* create rpc middleware file and add asic test

* move main to cmd package and have rpc package as an actual server

* move otel logic to rpc server

* add sanity checks for http calls

* fix review and make conversion extremely safe

* prevent infinite wait if tx has been dropped or does not exists

TXs can be dropped either because they are expired or the solver could not solve them. In that instance, we do not have to wait because all we do while waiting is to fetch the status of the tx on chain. If tx was dropped, then it cannot be onchain
